x-pack/plugins/stack_alerts/public/alert_types/es_query/expression.tsx

```diff
@@ -0,0 +1,368 @@
+    getDefaultParams()
+  );
+  const [testQueryResult, setTestQueryResult] = useState<string | null>(null);
+  const [testQueryError, setTestQueryError] = useState<string | null>(null);
+
+  const hasExpressionErrors = !!Object.keys(errors).find(
+    (errorKey) =>
+      expressionFieldsWithValidation.includes(errorKey) &&
+      errors[errorKey].length >= 1 &&
+      alertParams[errorKey as keyof EsQueryAlertParams] !== undefined
+  );
+
+  const expressionErrorMessage = i18n.translate(
+    'xpack.stackAlerts.esQuery.ui.alertParams.fixErrorInExpressionBelowValidationMessage',
+    {
+      defaultMessage: 'Expression contains errors.',
+    }
+  );
+
+  const setDefaultExpressionValues = async () => {
+    setAlertProperty('params', getDefaultParams());
+
+    setXJson(esQuery ?? DEFAULT_VALUES.QUERY);
+
+    if (index && index.length > 0) {
+      await refreshEsFields();
+    }
+  };
+
+  const setParam = (paramField: string, paramValue: unknown) => {
+    setCurrentAlertParams({
+      ...currentAlertParams,
+      [paramField]: paramValue,
+    });
+    setAlertParams(paramField, paramValue);
+  };
+
+  useEffect(() => {
+    setDefaultExpressionValues();
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  const refreshEsFields = async () => {
+    if (index) {
+      const currentEsFields = await getFields(http, index);
+      setEsFields(currentEsFields);
+    }
+  };
+
+  const hasValidationErrors = () => {
+    const { errors: validationErrors } = validateExpression(currentAlertParams);
+    return Object.keys(validationErrors).some(
+      (key) => validationErrors[key] && validationErrors[key].length
+    );
+  };
+
+  const onTestQuery = async () => {
+    if (!hasValidationErrors()) {
+      setTestQueryError(null);
+      setTestQueryResult(null);
+      try {
+        const window = `${timeWindowSize}${timeWindowUnit}`;
+        const timeWindow = parseDuration(window);
+        const parsedQuery = JSON.parse(esQuery);
+        const now = Date.now();
+        const { rawResponse } = await data.search
+          .search({
+            params: buildSortedEventsQuery({
+              index,
+              from: new Date(now - timeWindow).toISOString(),
+              to: new Date(now).toISOString(),
+              filter: parsedQuery.query,
+              size: 0,
+              searchAfterSortId: undefined,
+              timeField: timeField ? timeField : '',
+              track_total_hits: true,
+            }),
+          })
+          .toPromise();
+
+        const hits = rawResponse.hits;
+        setTestQueryResult(
+          i18n.translate('xpack.stackAlerts.esQuery.ui.numQueryMatchesText', {
+            defaultMessage: 'Query matched {count} documents in the last {window}.',
+            values: { count: hits.total, window },
+          })
+        );
+      } catch (err) {
+        const message = err?.body?.message;
```

Comment:
Looks like we're not including the root cause!
Luckily the data plugin exposes this under `err.body.attributes.error.root_cause[0].reason` so we can provide much more detailed error messages to the users - seems worth including this when we can. :)

Reply:
Awesome! I didn't know that. Updated to use that.